acceptance/indy_acceptance/indy-tests/test_scripts/acceptance_tests/remove_and_add_role.py

```diff
@@ -0,0 +1,421 @@
+"""
+Created on Nov 8, 2017
+
+@author: nhan.nguyen
+
+Containing test script of test scenario 09: remove and add role.
+"""
+import json
+from indy import ledger, signus
+from libraries.constant import Constant, Colors, Roles
+from libraries.result import Status
+from libraries.common import Common
+from libraries import utils
+from test_scripts.test_scenario_base import TestScenarioBase
+
+
+class TestScenario09(TestScenarioBase):
+
+    async def execute_test_steps(self):
+        """
+        This function is the main part of test script.
+        All steps that involve to role TGB (9, 10, a half of 24) will be skipped because
+        role TGB is not supported by libindy.
+        There is a bug in this scenario (in step 22, 23 24) so we log a bug here.
+        """
+        try:
+            # 1. Create and open wallet, pool ledger.
+            self.steps.add_step("Create and open wallet, pool ledger")
+            result = await utils.perform(self.steps, Common.prepare_pool_and_wallet,
+                                         self.pool_name, self.wallet_name, Constant.pool_genesis_txn_file)
+            utils.exit_if_exception(result)
+            (self.pool_handle, self.wallet_handle) = result
+
+            # 2. Create DIDs.
+            self.steps.add_step("Create DIDs")
+
+            result = await utils.perform(self.steps, signus.create_and_store_my_did,
+                                         self.wallet_handle, json.dumps({"seed": Constant.seed_default_trustee}))
+            default_trustee_did = result[0] if len(result) == 2 else (None, None)
+
+            result = await utils.perform(self.steps, signus.create_and_store_my_did,
```

Comment:
The status of the current step is overwritten many times. So FAILED status may be overwritten by PASSED that is obviously not the expected behavior.

Reply:
This issue has not been fixed.

Reply:
Please refer to 
https://github.com/trongnhan1312400/indy-node/blob/84876afddf1eb493652ca089e8449e9325c0a3eb/acceptance/indy_acceptance/test_scripts/acceptance_tests/remove_and_add_role.py 
for the fixed code.

Reply:
FYI. It was uploaded to "Commits on Dec 19, 2017 "

Reply:
In step 2 "Create DIDs" multiple actions are done by calls of `utils.perform`. Each call of `utils.perform` here is done with `ignore_exception=True` which is used by default. So, for example, if creation of trustee1 fails then the current step status will be set to FAILED. But next if creation of trustee2 succeeds then the current step status is overwritten from FAILED to PASSED. Such the behavior is incorrect. Obviously, if some action in the step has failed then the whole step status must be FAILED.

To correct this behavior I propose to add every new step with PASSED status initially and don't update the step's status in `utils.perform` in case `func` has succeeded. With this logic we will get the step passed only if all the actions in it have succeed and we will get the whole step failed if any action in it has failed.

By the way, it is ok to add a new step with the initial PASSED status because if some error occurred not inside `utils.perform` call aborts a test case then the step's status will be set to FAILED in `except` section of `try` block in `TestScenarioBase.execute_scenario`.

Reply:
As some best practices, the default status of step should be FAILED. With ignore_exception=True, utils.perform will update the status after done as 2 options below:
-If the action is PASSED, the status of step will be updated as PASSED.
-If the action is FAILED, the status of step will be updated to FAILED. Then we freeze this status from other updates. That means we won't change this step's status anymore.

Please refer to the latest code for detail.

Reply:
Just noticed the freeze logic in `Step.set_status`. It has fixed the issue.